src/conn/conn_dhandle.c

```diff
@@ -772,7 +781,7 @@ __conn_dhandle_close_one(
         if (ret == 0)
             ret = __wt_meta_track_sub_off(session);
     }
-    if (removed)
+    if (removed && (ret != EBUSY))
```

Comment:
We have already returned in the added check. Why this is needed?

Reply:
This is needed to skip the next line F_SET(session->dhandle, WT_DHANDLE_DROPPED); so the drop operation is a NOP if EBUSY is returned. Secondly, __session_drop() ends with API_END_RET_NOTFOUND_MAP_OK(session, ret, EBUSY); instead of API_END_RET_NOTFOUND_MAP(session, ret); so if EBUSY session->txn is not marked an error by __wt_trn_err_set(). Without this change session->txn would need to be rolled back. These two changes make __session_drop() a NOP if EBUSY is returned here. Thus __session_drop() can be retired and will work if EBUSY is not returned again.

Reply:
Instead of checking EBUSY, can we generalize it to check ret != 0?

Reply:
Generalizing to ret !=0 has a few problems. (1) If ret != EBUSY the error might not be a NOP, i.e. something might have been done despite the error. (2) There are two changes to make drop_table() a NOP for EBUSY. It would also be necessary to change __session_drop() to accept other error codes too.

Reply:
What happens if removed is true and ret is EBUSY? I understand that we want to turn this into a NOP if we return EBUSY but does this mean we still have removed something in this case and it is not a NOP?

Would it possible to assert that if we have removed something, we are not getting EBUSY?

Reply:
Instead of relying on the error code, is it possible to pass in a boolean pointer to __wt_conn_dhandle_close to indicate if we have done any real work? Something like this?

if (is_btree) {
        /*
         * !! Check if we have exclusive access here, else a transaction can start on this btree
         * after this check !!
         *
         * Don't close the btree if there are changes not yet globally visible.
         */
        WT_ASSERT_ALWAYS(session, btree->max_upd_txn != WT_TXN_ABORTED,
          "Assert failure: session: %s: btree->max_upd_txn == WT_TXN_ABORTED", session->name);
        if (check_visibility && !__wt_txn_visible_all(session, btree->max_upd_txn, WT_TS_NONE))
               return (EBUSY)
        
        *do_real_work = true;
  }
  
  I think in this way it is clear for the caller that it hasn't done anything on the dhandle and we don't introduce the assumption that EBUSY means we haven't done any real work but other error code might.

Reply:
In discussion https://wiki.corp.mongodb.com/pages/viewpage.action?pageId=268092082 I found out that WiredTiger has a policy: error WT_NOTFOUND does not require a transaction to be rolled back, but every other error does. Making drop table EBUSY be a NOP violates this policy. Thus Alex recommended I leave out the change in __ session_drop() to not force transaction rollback, and include the change in __ wt_conn_dhandle_close() to not set WT_DHANDLE_DROPPED on EBUSY. Since it is not possible to prevent a partial drop table of complex tables by __ wt_conn_dhandle_close(), Alex recommended I disallow drop table force=true returning ENOTSUP for complex tables. That change is in __ drop_table().

Given the above policy, the assumption that EBUSY means we haven't done any real work is gone and there is no need to return via *do_real_work.

Reply:
ok, understood.

Reply:
Can we add a comment to explain why ebusy is special here?

Reply:
I don't know what you want. EBUSY was going to be special. Then because of the policy EBUSY is no longer special. That change was removed.

Reply:
If EBUSY is no longer special, can we generalize that the check to ret != 0? But based on your previous comment, we cannot do that. I would like the comment to explain why we cannot just do ret != 0.

Reply:
Changed the if on new line 784 back to old line 775.